FP8 ACLE specification (#323)
ARM-software/acle: main/acle.md

Providing emulation libraries for half-precision floating point
conversions when not implemented in hardware is implementation-defined.

### Modal 8-bit floating-point

ACLE defines the `__mfp8` type, which can be used for the E5M2 and E4M3

> 8-bit floating-point formats. It is a storage and interchange only type

Author: done

ACLE intrinsics are available. This implies that `__ARM_FEATURE_SM4` and
`__ARM_FEATURE_SVE2` are both nonzero.

> ### Lookup table extensions

Reviewer: Looks like this section should be part of a separate PR.

Author: Ack, it for/in https://github.com/ARM-software/acle/pull/324

lookup table instructions with 4-bit indices and 8-bit elements (FEAT_SME_LUTv2)
and if the associated ACLE intrinsics are available.

### Modal 8-bit floating point extensions

> `__ARM_FEATURE_FP8` is defined to 1 if there is hardware support for FP8 conversion

Reviewer 2: set to

> `__ARM_FEATURE_FP8` is defined to 1 if there is hardware support for FP8 conversion

Reviewer 2: set to

Reviewer 2: In change 364, 1 is wrapped with `1`

Bit patterns other than as described above are invalid. Passing an invalid value as an argument
to an FP8 intrinsic results in undefined behavior.

The ACLE declares several helper types and intrinsics to
facilitate construction of `fpm` arguments. The helper intrinsics do not have
side effects and their return depends only on their parameters.

The helper types and intrinsics are available after including any of
[`<arm_neon.h>`](#arm_neon.h), [`<arm_sve.h>`](#arm_sve.h), or
[`<arm_sme.h>`](#arm_sme.h).

Note: where a helper intrinsic description refers to "updating the FP8 mode" it
means the intrinsic only modifies the bits of the input `fpm_t` parameter that
correspond to the new mode and returns the resulting value. No side effects
(such as changing processor state) occur.

## Support enumerations

```c
enum __ARM_FPM_FORMAT {
  __ARM_FPM_E5M2,
  __ARM_FPM_E4M3,
};

enum __ARM_FPM_OVERFLOW {
  __ARM_FPM_INFNAN,
  __ARM_FPM_SATURATE,
};
```

## Helper intrinsics

```c
  fpm_t __arm_fpm_init();
```
Initializes a value, suitable for use as an `fpm` argument ("FP8 mode").
The value corresponds to a mode of operation where:

>   * the sources and destination operands are interpreted as E5M2

Author: done

```c
  fpm_t __arm_set_fpm_overflow_cvt(fpm_t fpm, enum __ARM_FPM_OVERFLOW behavior);
```
Updates the FP8 mode to set the overflow behavior for conversion operations.

``` c
  fpm_t __arm_set_fpm_overflow_mul(fpm_t fpm, enum __ARM_FPM_OVERFLOW behavior);
```
Updates the FP8 mode to set the overflow behavior for multiplicative
operations.

``` c
  fpm_t __arm_set_fpm_lscale(fpm_t fpm, uint64_t scale);
```
Updates the FP8 mode to set the downscaling value that is subtracted from:
* The product or the sum-of-products exponent, for multiplication instructions
   with FP8 operands.
* The result exponent, for instructions that convert the first FP8
  input data stream to other floating-point formats.

The valid range for the `scale` parameter is [0, 127], inclusive.

``` c
  fpm_t __arm_set_fpm_lscale2(fpm_t fpm, uint64_t scale);
```
Updates the FP8 mode to set the downscaling value that is subtracted from
the result exponent for instructions that convert
the second FP8 input data stream to other floating-point formats.

The valid range for the `scale` parameter is [0, 63], inclusive.

``` c
  fpm_t __arm_set_fpm_nscale(fpm_t fpm,  int64_t scale);
```

> Updates the FP8 mode valiue to set the scaling value that is added to

Author: Done

| 24-31         | `nscale`       | scaling value for conversions                                      |
| 32-37         | `lscale2`      | downscaling value for conversions of the second input stream       |
| 38-63         |                | must be zero                                                       |

Bit patterns other than as described above are invalid. Passing an invalid value as an argument
to an FP8 intrinsic results in undefined behavior.

The ACLE declares several helper types and intrisics to
facilitate construction of `fpm` arguments. The helper intrinsics do not have
side effects and their return depends only on their parameters.

Note: where a helper intrinsic description refers to "updating the FP8 mode"
it means the intrinsic only modifies the bits of input `fpm_t` parameter that
correspond to the new mode and returns the resulting value. No side effects
(such as changing processor state) occur.

## Support enumerations

```c
enum __ARM_FPM_FORMAT {
  __ARM_FPM_E5M2,
  __ARM_FPM_E4M3,
};

enum __ARM_FPM_OVERFLOW {
  __ARM_FPM_INFNAN,
  __ARM_FPM_SATURATE,
};
```

## Helper intrinsics

```c
  fpm_t __arm_fpm_init();

> ```

Reviewer 3: Can you add some wording saying which header makes these definitions available please?
From what I can tell the proposed LLVM patch (https://github.com/llvm/llvm-project/pull/100608) defines them in arm_neon.h but the GCC patch (https://gcc.gnu.org/pipermail/gcc-patches/2024-July/658262.html) defines them arm_acle.h

# About FP8 intrinsics

The specification for FP8 intrinsics is in
[**Alpha** state](#current-status-and-anticipated-changes).

Each 8-bit floating point intrinsic call has a parameter to define the format,
saturation, and scale of its operands. This parameter is typically
declared as `fpm_t fpm`.

```c
  typedef uint64_t fpm_t;
```

The ACLE declares several helper types and functions to
facilitate construction of `fpm` arguments.

## Support enumerations

```c
enum __ARM_FPM_FORMAT {
  __ARM_FPM_E5M2,
  __ARM_FPM_E4M3,
};

enum __ARM_FPM_OVERFLOW {
  __ARM_FPM_INFNAN,
  __ARM_FPM_SATURATE,
};
```

## Helper functions

```c
  fpm_t arm_fpm_init();

  fpm_t arm_set_fpm_src1_format(fpm_t fpm, enum __ARM_FPM_FORMAT format);
  fpm_t arm_set_fpm_src2_format(fpm_t fpm, enum __ARM_FPM_FORMAT format);
  fpm_t arm_set_fpm_dst_format(fpm_t fpm,  enum __ARM_FPM_FORMAT format);

  fpm_t arm_set_fpm_overflow_cvt(fpm_t fpm, enum __ARM_FPM_OVERFLOW behaviour);
  fpm_t arm_set_fpm_overflow_mul(fpm_t fpm, enum __ARM_FPM_OVERFLOW behaviour);

  fpm_t arm_set_fpm_lscale(fpm_t fpm, uint64_t scale);
  fpm_t arm_set_fpm_lscale2(fpm_t fpm, uint64_t scale);

>   fpm_t arm_set_fpm_nscale(fpm_t fpm,  int64_t scale);

Reviewer 4: How should scale be interpreted? we have a specific number of bits, what should the behaviour be if the value provided is higher than what can fit in them?

>   fpm_t arm_set_fpm_nscale(fpm_t fpm,  int64_t scale);

Reviewer 4: How about adding this documentation? 
```
  /* Returns a default value of fpm. It is unspecified whether the returned value
     is valid. Users must call the helper functions below, using the output of
     each as a parameter to the next, in order to obtain a valid floating
     point mode value. */
  fpm_t arm_fpm_init();

  /* Updates the first FP8 input data stream format. */
  fpm_t arm_set_fpm_src1_format(fpm_t fpm, enum __ARM_FPM_FORMAT format);

  /* Updates the second FP8 input data stream format. */
  fpm_t arm_set_fpm_src2_format(fpm_t fpm, enum __ARM_FPM_FORMAT format);

  /* Updates the destination data stream format. */
  fpm_t arm_set_fpm_dst_format(fpm_t fpm,  enum __ARM_FPM_FORMAT format);

  /* Updates the result when a floating-point Overflow exception is
     detected during FP8 convert instructions. */
  fpm_t arm_set_fpm_overflow_cvt(fpm_t fpm, enum __ARM_FPM_OVERFLOW behaviour);

  /* Updates the result when a floating-point Overflow exception is
     detected during FP8 multiplication instructions. */
  fpm_t arm_set_fpm_overflow_mul(fpm_t fpm, enum __ARM_FPM_OVERFLOW behaviour);

  /* Updates the downscaling value. This value is an unsigned integer that
     is subtracted from:
     - The product or the sum-of-products exponent, for multiplication
       instructions with FP8 operands.
     - The result exponent, for instructions that convert the first FP8
       input data stream to other floating-point formats.
     Valid values are 0-127. The result of using values larger than 127 is
     unspecified. */
  fpm_t arm_set_fpm_lscale(fpm_t fpm, uint64_t scale);
  
  /* Updates the downscaling value for instructions that convert the second
     FP8 input data stream to other floating-point formats. This value is an
     unsigned integer that is subtracted from the result exponent.
     Valid values are 0-63. The result of using values larger than 63 is
     unspecified. */
  fpm_t arm_set_fpm_lscale2(fpm_t fpm, uint64_t scale);
  
  /* Updates the scaling value for instructions that convert other
     floating-point formats to an FP8 format. This value is a signed integer
     that is added to the operand exponent.
     Valid values are -128 to 127. The result of using values outside this
     range is unspecified. */
  fpm_t arm_set_fpm_nscale(fpm_t fpm,  int64_t scale);
```

Author: Some descriptions added.

> ```

Reviewer 5: These need a description of what they need to do.  It's not enough just to add some prototypes.  What are the constraints on 'scale' for the last three?  What does arm_fpm_init() do?
Do these need to be real functions (eg, can I take the address of one)?  If so, where should they be defined in the build system?  Should the functions be marked as __attribute__((const)) or equivalent in the new attribute syntax, to indicate that they have no side effects (or if that's too much, 'pure')?

Author: I added descriptions.

> Do these need to be real functions (eg, can I take the address of one)? If so, where should they be defined in the build system?

I've reworded the text to refer to them as "intrinsics" and as such they fall into this category:

> Whether intrinsics are macros, functions or built-in operators is
> unspecified. 
> 

> Should the functions be marked as attribute((const)) or equivalent in the new attribute syntax, to indicate that they have no side effects (or if that's too much, 'pure')?

I've added clarifying text that these helper intrinsics do not have side effects and their result depend only on the arguments.  It's up to the implementations to choose whether and how to represent this information and choose a particular syntax.

|               |                |   0 - generate infinity, 1 - generate maxumum normal number        |
| 15            | `overflow_cvt` | overflow behavior for conversion inrtructions:                     |
|               |                |   0 - generate infinity or NaN, 1 - generate maximum normal number |
| 16-22         | `lscale`       | downscaling value                                                  |
| 23            |                | must be zero                                                       |
| 24-31         | `nscale`       | scaling value for conversions                                      |
| 32-37         | `lscale2`      | downscaling value for conversions of the second input stream       |
| 38-63         |                | must be zero                                                       |

Bit patterns other than as described above are invalid. Passing an invalid value as an argument
to an FP8 intrinsic results in undefined behavior.

The ACLE declares several helper types and intrinsics to
facilitate construction of `fpm` arguments. The helper intrinsics do not have

> side effects and their return depends only on their parameters.

Reviewer: ```suggestion
side effects and their return values depend only on their parameters.
```

The helper types and intrinsics are available after including any of
[`<arm_neon.h>`](#arm_neon.h), [`<arm_sve.h>`](#arm_sve.h), or
[`<arm_sme.h>`](#arm_sme.h).

Note: where a helper intrinsic description refers to "updating the FP8 mode" it
means the intrinsic only modifies the bits of the input `fpm_t` parameter that

> correspond to the new mode and returns the resulting value. No side effects

Author: No, it's the "bits" that correspond to the new mode.

(such as changing processor state) occur.

## Support enumerations

```c
enum __ARM_FPM_FORMAT {
  __ARM_FPM_E5M2,
  __ARM_FPM_E4M3,
};

enum __ARM_FPM_OVERFLOW {
  __ARM_FPM_INFNAN,
  __ARM_FPM_SATURATE,
};
```

## Helper intrinsics

```c
  fpm_t __arm_fpm_init();
```
Initializes a value, suitable for use as an `fpm` argument ("FP8 mode").
The value corresponds to a mode of operation where:

>   * The sources and destination operands are interpreted as E5M2.

Reviewer: ```suggestion
  * The source and destination operands are interpreted as E5M2.
```

```
Updates the FP8 mode to set the overflow behavior for conversion operations.

``` c
  fpm_t __arm_set_fpm_overflow_mul(fpm_t fpm, enum __ARM_FPM_OVERFLOW behavior);
```
Updates the FP8 mode to set the overflow behavior for multiplicative
operations.

``` c
  fpm_t __arm_set_fpm_lscale(fpm_t fpm, uint64_t scale);
```
Updates the FP8 mode to set the downscaling value subtracted from:
* The product or the sum-of-products exponent, for multiplication instructions
  with FP8 operands.
* The result exponent, for instructions converting the first FP8
  input data stream to other floating-point formats.

The valid range for the `scale` parameter is [0, 127], inclusive.

``` c
  fpm_t __arm_set_fpm_lscale2(fpm_t fpm, uint64_t scale);
```
Updates the FP8 mode to set the downscaling value subtracted from the
result exponent for instructions converting the second FP8 input data
stream to other floating-point formats.

The valid range for the `scale` parameter is [0, 63], inclusive.

``` c
  fpm_t __arm_set_fpm_nscale(fpm_t fpm,  int64_t scale);
```
Updates the FP8 mode to set the scaling value added to the operand's

> exponent for instructions converting other floating-point formats to a

Reviewer: ```suggestion
exponent for instructions converting other floating-point formats to an
```

FP8 format.

The valid range for the `scale` parameter is [-128, 127], inclusive.

> Passing an out of range argument to a helper intrinsic results in the intrinsic

Reviewer: I think this should be placed before `## Support enumerations`.  As currently presented and formatted, it looks like it's part of the `__arm_set_fpm_nscale(` description.

| 6-8           | `dst_format`   | destination format: 0  - FP8 E5M2, 1 - FP8 E4M3                    |
| 9-13          |                | must be zero                                                       |
| 14            | `overflow_mul` | overflow behavior for multiplication instructions:                 |
|               |                |   0 - generate infinity, 1 - generate maximum normal number        |
| 15            | `overflow_cvt` | overflow behavior for conversion instructions:                     |
|               |                |   0 - generate infinity or NaN, 1 - generate maximum normal number |
| 16-22         | `lscale`       | downscaling value                                                  |
| 23            |                | must be zero                                                       |
| 24-31         | `nscale`       | scaling value for conversions                                      |
| 32-37         | `lscale2`      | downscaling value for conversions of the second input stream       |
| 38-63         |                | must be zero                                                       |

> Bit patterns other than as described above are invalid. Passing an invalid value as an argument

Reviewer 2: Bit patterns other than those described in this table are invalid.

to an FP8 intrinsic results in undefined behavior.

The ACLE declares several helper types and intrinsics to
facilitate construction of `fpm` arguments. The helper intrinsics do not have
side effects and their return values depend only on their parameters.

Passing an out of range argument to a helper intrinsic results in the intrinsic
returning an indeterminate value. Passing such an indeterminate value as
an argument to an FP8 intrinsic results in undefined behavior.

The helper types and intrinsics are available after including any of
[`<arm_neon.h>`](#arm_neon.h), [`<arm_sve.h>`](#arm_sve.h), or
[`<arm_sme.h>`](#arm_sme.h).

> Note: where a helper intrinsic description refers to "updating the FP8 mode" it

Reviewer 2: Remove the quotes around "updating the FP8 mode"

means the intrinsic only modifies the bits of the input `fpm_t` parameter that
correspond to the new mode and returns the resulting value. No side effects

> (such as changing processor state) occur.

Reviewer 2: , such as changing processor state, occur.

Individual FP8 intrinsics are described in their respective
Advanced SIMD (NEON), SVE, and SME sections.

## Support enumerations

```c
enum __ARM_FPM_FORMAT {
  __ARM_FPM_E5M2,
  __ARM_FPM_E4M3,
};

enum __ARM_FPM_OVERFLOW {
  __ARM_FPM_INFNAN,
  __ARM_FPM_SATURATE,
};
```

## Helper intrinsics

```c
  fpm_t __arm_fpm_init();
```

> Initializes a value, suitable for use as an `fpm` argument ("FP8 mode").

Reviewer 2: in FP8 mode. ?

The value corresponds to a mode of operation where:
  * The source and destination operands are interpreted as E5M2.

>   * Overflow behavior is to yield infinity or NaN (depending on operation).

Reviewer 2: , depending on operation.

### Vector data types

Vector data types are named as a lane type and a multiple. Lane type

> names are based on the types defined in `<stdint.h>`. For example,.

Author: done

`int16x4_t` is a vector of four `int16_t` values. The base types are
`int8_t`, `uint8_t`, `int16_t`, `uint16_t`, `int32_t`, `uint32_t`,
`int64_t`, `uint64_t`, `float16_t`, `float32_t`, `poly8_t`, `poly16_t`,

> `poly64_t`, `poly128_t` , `bfloat16_t`. The multiples are such that the

Author: done

when supported by the hardware. The bfloat types are all opaque types.  That is
to say they can only be used by intrinsics.

FP8 types are only available when the `__mfp8` type is defined, i.e.

> when supported by the hardware. The FP8 types are all opaque types.  That is

Author: done

> when supported by the hardware. The bfloat types are all opaque types.  That is

Reviewer 2: That is,
they can only be used by intrinsics.

to say they can only be used by intrinsics.

> The FP8 types are all opaque types. That is to say they can only be used

Reviewer 2: That is, they can only be used

> by intrinsics.

Reviewer 6: That is to say they can only be used by intrinsics.

Change to: That is, they can only be used by intrinsics.

| **Signed integer**   | **Unsigned integer** | **Floating-point**   |                      |
| -------------------- | -------------------- | -------------------- | -------------------- |

> | `svint8_t`           | `svuint8_t`          |                      | `svmfloat8_t         |

Reviewer 4: Should the 

> `svbfloat16_t` is only available if the header file also provides a definition of `bfloat16_t`  

section below be repeated or extended for svmfloat8_t?

Author: Wouldn't that be necessary only if the underlying type definition was conditional?

Reviewer 4: It would, I was confused, thanks for clarifying :)

``` c
   // Variants are also available for:
   // _u8, _u16, _s16, _u32, _s32, _u64, _s64

>    // _mf8_, _bf16, _f16, _f32, _f64

Author: done

   // _u16, _u32, _s32, _u64, _s64
   // _bf16, _f16, _f32, _f64
   svint16_t svld1q_gather_[u64]index[_s16](svbool_t pg, const int16_t *base, svuint64_t index);

>    svint8_t svld1q_gather[_u64base]_index_s8(svbool_t pg, svuint64_t zn, int64_t index);

Author: Do you like this?
```
svint16_t svld1q_gather[_u64base]_index_s16(svbool_t pg, svuint64_t zn, int64_t offset);
```

>    svint8_t svld1q_gather[_u64base]_index_s8(svbool_t pg, svuint64_t zn, int64_t index);

Reviewer: Yeah, looks good to me.

>    svint8_t svld1q_gather[_u64base]_index_s8(svbool_t pg, svuint64_t zn, int64_t index);

Author: I've added those changes to https://github.com/ARM-software/acle/pull/337

``` c
   // Variants are also available for:
   // _u8, _u16, _s16, _u32, _s32, _u64, _s64
   // _mf8, _bf16, _f16, _f32, _f64
   svint8_t svld1q_gather[_u64base]_s8(svbool_t pg, svuint64_t zn);
   svint8_t svld1q_gather[_u64base]_offset_s8(svbool_t pg, svuint64_t zn, int64_t offset);

>    svint8_t svld1q_gather[_u64]_offset[_s8](svbool_t pg, const int8_t *base, svuint64_t offset);

Reviewer: The original `_[u64]offset` suffix was correct.  Same for the suffix changes in the gather and scatter functions below.

Author: Ack. Please, have a look at https://github.com/ARM-software/acle/pull/337

   svint8_t svld1q_gather_[u64]offset[_s8](svbool_t pg, const int8_t *base, svuint64_t offset);
   svint8_t svld1q_gather[_u64]_offset[_s8](svbool_t pg, const int8_t *base, svuint64_t offset);

   // Variants are also available for:
   // _u16, _u32, _s32, _u64, _s64
   // _bf16, _f16, _f32, _f64

>    svint16_t svld1q_gather_[u64]index[_s16](svbool_t pg, const int16_t *base, svuint64_t index);

Reviewer: We shouldn't delete this (AFAIK).  My comment below was specific to that one line.

>    svint16_t svld1q_gather[_u64]_index[_s16](svbool_t pg, const int16_t *base, svuint64_t index);

Reviewer: ```suggestion
   svint16_t svld1q_gather[_u64base]_index[_s16](svbool_t pg, const int16_t *base, svuint64_t index);
```

(The version in the resolved comment was correct.)

Author: No more gather/scatter changes in this patch.

``` c
   // Variants are also available for:
   // _u8, _u16, _s16, _u32, _s32, _u64, _s64
   // _bf16, _f16, _f32, _f64
   // _bf16, _f16, _f32, _f64, _mf8
   void svst1q_scatter[_u64base][_s8](svbool_t pg, svuint64_t zn, svint8_t data);
   void svst1q_scatter[_u64base]_offset[_s8](svbool_t pg, svuint64_t zn, int64_t offset, svint8_t data);
   void svst1q_scatter_[u64]offset[_s8](svbool_t pg, const uint8_t *base, svuint64_t offset, svint8_t data);

   // Variants are also available for:
   // _u16, _u32, _s32, _u64, _s64
   // _bf16, _f16, _f32, _f64

>    void svst1q_scatter[_u64base]_index[_s8](svbool_t pg, svuint64_t zn, int64_t index, svint8_t data);

Reviewer: The current patch still drops the function.

``` c
   // Variants are also available for:
   // _s8, _u16, _s16, _u32, _s32, _u64, _s64
   // _bf16, _f16, _f32, _f64, _mf8
   svuint8_t svzipq1[_u8](svuint8_t zn, svuint8_t zm);
   svuint8_t svzipq2[_u8](svuint8_t zn, svuint8_t zm);
   ```

> ### Modal 8-bit floating-point intrinsics

Author: Do you mean having FP8 as a separate subsection to `### SVE2.1 and SME2 instruction intrinsics` or just add the entries inline?
Likewise for `### SME2 modal 8-bit floating-point intrinsics`, does it stay as a separate subsection ?

> ### Modal 8-bit floating-point intrinsics

Author: FP8 instructions are SVE2, not SVE2.1, so maybe `### SVE2.1 and SME2 instruction ...` is not the right section.

> ### Modal 8-bit floating-point intrinsics

Reviewer: Yeah, I meant as a separate section, but alongside that one.  I think it would be clearer to keep SVE+SME intrinsics together.

#### BF1CVT, BF2CVT

8-bit floating-point convert to BFloat16.
``` c

>   svbfloat16_t svcvt1_bf16[_mf8]_fpm(svmfloat8_t zn, fpm_t fpm);

Author: I think `svcvtN_xxx` and `svcvtltN_xxx` look more like having a consistent naming with the non-FP8 `svcvt` and `svcvtlt` without `N` sticking in the middle and interrupting the flow.

```

### SVE2 and SME2 modal 8-bit floating-point intrinsics

The intrinsics in this section are defined by the header file
[`<arm_sve.h>`](#arm_sve.h) when `__ARM_FEATURE_FP8` is defined,
and `__ARM_FEATURE_SVE2` or `__ARM_FEATURE_SME2` is defined. Individual
intrinsics may have additional target feature requirements.

#### BF1CVT, BF2CVT, F1CVT, F2CVT

8-bit floating-point convert to half-precision and BFloat16.
``` c
  // Variants are also available for: _bf16
  svfloat16_t svcvt1_f16[_mf8]_fpm(svmfloat8_t zn, fpm_t fpm);
  svfloat16_t svcvt2_f16[_mf8]_fpm(svmfloat8_t zn, fpm_t fpm);
```

#### BF1CVTLT, BF2CVTLT, F1CVTLT, F2CVTLT

8-bit floating-point convert to half-precision and BFloat16 (top).
``` c
  // Variants are also available for: _bf16
  svfloat16_t svcvtlt1_f16[_mf8]_fpm(svmfloat8_t zn, fpm_t fpm);
  svfloat16_t svcvtlt2_f16[_mf8]_fpm(svmfloat8_t zn, fpm_t fpm);
```

#### BFCVTN, FCVTN

Half-precision and BFloat16 convert, narrow and interleave to 8-bit
floating-point.
``` c
  // Variant is also available for: _bf16_x2
  svmfloat8_t svcvtn_mf8[_f16_x2]_fpm(svfloat16x2_t zn, fpm_t fpm);
```

#### FCVTNT, FCVTNB

> Single-precision convert, narrow and interleave to 8-bit floating-point (top and bottom).

Reviewer 6: comma after 'narrow'
